packages/react-dom/src/client/ReactDOMComponent.js

```diff
@@ -264,14 +264,31 @@ if (__DEV__) {
   };
 }
 
+/**
+ * If the root container is inside of a ShadowRoot, events will lost.
+ * Recursively access its' parentNode to check if it's inside a ShadowRoot.
+ *
+ * If developer mount the node to the dom first,
+ * then attach it to a ShadowRoot, events will still lost.
+ */
+function isInDocumentFragment(element?: Element | Node | null) {
+  if (!element) {
+    return false;
+  }
+  if (element.nodeType === DOCUMENT_FRAGMENT_NODE) {
+    return true;
+  }
+  return isInDocumentFragment(element.parentNode);
```

Comment:
It's better to do this in a while loop rather than recursively.

Reply:
hmm, a dom has a limited set of parents so the recursive chain won't be too long. And this recursive function doesn't introduce any new variables in itself and have clear quit condition so the recursive is safe I think

Reply:
A DOM node can have a huge DOM tree, with thousands of parents. I still feel we should use a while loop here, as the extra bytes will be neglible.